### PR TITLE
Change url to Loggly Gen2, adding support to Tags, Exception handling

### PR DESCRIPTION
**Pull request by psquickitjayant**
Deleted old repository due to lots of conflicts. Merged all the changes in the new branch. 

Changes include

<ol>
<li>Updated README to add a link to Loggly docs</li>
<li>Updated Loggly url to Gen2 account</li>
<li>Added support for tags</li>
<li>Supports dynamic tag values. thanks @nywilken</li>
<li>Added Exception Handling</li>
<li>Retires up to five times</li>
</ol>

### CODE REVIEW

**Comment by mostlyjason** [comment]
Nice thanks for merging these all together. Hopefully it'll make it easier for them to accept the PR.

File: lib/logstash/outputs/loggly.rb [hunk 1]
```diff
@@ -43,52 +43,73 @@ class LogStash::Outputs::Loggly < LogStash::Outputs::Base
...
-    end
-    request = Net::HTTP::Post.new(url.path)
-    request.body = event.to_json
-    response = http.request(request)
-    if response.is_a?(Net::HTTPSuccess)
-      @logger.info("Event send to Loggly OK!")
-    else
-      @logger.warn("HTTP error", :error => response.error!)
-    end
+    5.times do #retries up to five times
```
Comment (suyograo) [inline lib/logstash/outputs/loggly.rb]: can you provide this as an option? and not hardcoded?

Comment (suyograo) [inline lib/logstash/outputs/loggly.rb]: maybe use https://github.com/jordansissel/ruby-stud/blob/master/lib/stud/try.rb

**Comment by suyograo** [comment]
@psquickitjayant @nywilken So can we close https://github.com/logstash-plugins/logstash-output-loggly/pull/5 in lieu of this?

**Comment by nywilken** [comment]
@psquickitjayant @suyograo PR https://github.com/logstash-plugins/logstash-output-loggly/pull/6 does not contain all of the changes submitted in PR https://github.com/logstash-plugins/logstash-output-loggly/pull/5 so we should not close it just yet. 

@psquickitjayant your updated PR does not contain the latest plugin refactor made for testing and error handling -- see [c561f069](https://github.com/logstash-plugins/logstash-output-loggly/pull/5/files#diff-c561f06941a5987d4b5f9c26d87deb1bR89)

@suyograo If PR 5 is merged would it be easier to have @psquickitjayant sync his fork with the updated upstream then update PR 6? Or should I submit my PR to @psquickitjayant and close PR 5?

thanks

**Comment by nywilken** [comment]
@psquickitjayant can you explain the use case for the exception handling and retry logic; More specifically, could you explain which exceptions you are interested in retrying on? I think the scope of the begin/rescue is a bit too broad. 

The rescue will catch and retry on any exception that is thrown within the receive method. For exceptions that indicate a hard failure (Net::HTTPBadRequest, Net::HTTPBadGateway, etc.) a retry should not be executed as they will result in the same exception. Is it possible to explicitly specify the exceptions you wish to catch? 

As for the retry logic I agree with @suyograo that it should be a config option. I would also request that the retry only be executed on the post request to Loggly--unless you have a different use case. Do you know if the Loggly API has any documentation on client retries? 

For exception handling I generally try to follow the practices outlined on the [Ruby-Style Guide] (https://www.relishapp.com/womply/ruby-style-guide/docs/exceptions). 

Let me know if there is anything I can do to help with getting the changes from PR 5 merged into your changes or with the exception handling logic.

**Comment by mostlyjason** [comment]
+1 on making it configurable. Here is a comment from a customer on two types of exceptions. We should rescue and retry on these if we can "two things I've noticed from cases is that the plugin will crash after a 502 "Bad Gateway" or Logstash will throw Timeout::Error: execution expired resulting in a crash as well."

**Comment by psquickitjayant** [comment]
@suyograo @nywilken  I am updating plugin to make it configurable. Retry mode will work by default unless user explicitly configures it to No / false

I think we need to add rescue for both Bad Gateway and Bad Request to prevent Logstash from crashing as mentioned by @mostlyjason 

Closing this PR for now.

**Comment by nywilken** [comment]
@psquickitjayant thanks for your work on updating the the exception handling logic. The proposed changes sound good. While you refactor the logic in this PR can we ask that PR https://github.com/logstash-plugins/logstash-output-loggly/pull/5 be merged?

**Comment by psquickitjayant** [comment]
@nywilken Sure, if @suyograo is fine with that. Then I can take the latest changes and update my code accordingly.

**Pull request closed without merging**